src/crimson/mgr/client.cc

```diff
@@ -0,0 +1,129 @@
+#include "client.h"
+
+#include "crimson/common/log.h"
+#include "crimson/net/Connection.h"
+#include "crimson/net/Messenger.h"
+#include "messages/MMgrConfigure.h"
+#include "messages/MMgrMap.h"
+#include "messages/MMgrOpen.h"
+
+namespace {
+  seastar::logger& logger()
+  {
+    return ceph::get_logger(ceph_subsys_mgrc);
+  }
+  template<typename Message, typename... Args>
+  Ref<Message> make_message(Args&&... args)
+  {
+    return {new Message{std::forward<Args>(args)...}, false};
+  }
+}
+
+using ceph::common::local_conf;
+
+namespace ceph::mgr
+{
+
+Client::Client(ceph::net::Messenger& msgr,
+                 WithStats& with_stats)
+  : msgr{msgr},
+    with_stats{with_stats},
+    report_timer{[this] {report();}}
+{}
+
+seastar::future<> Client::start()
+{
+  return seastar::now();
+}
+
+seastar::future<> Client::stop()
+{
+  return gate.close().then([this] {
+    if (conn) {
+      return conn->close();
+    } else {
+      return seastar::now();
+    }
+  });
+}
+
+seastar::future<> Client::ms_dispatch(ceph::net::ConnectionRef conn,
+                                      MessageRef m)
+{
+  switch(m->get_type()) {
+  case MSG_MGR_MAP:
+    return handle_mgr_map(conn, boost::static_pointer_cast<MMgrMap>(m));
+  case MSG_MGR_CONFIGURE:
+    return handle_mgr_conf(conn, boost::static_pointer_cast<MMgrConfigure>(m));
+  default:
+    return seastar::now();
+  }
+}
+
+seastar::future<> Client::ms_handle_reset(ceph::net::ConnectionRef c)
+{
+  if (conn == c) {
+    return reconnect();
+  } else {
+    return seastar::now();
+  }
+}
+
+seastar::future<> Client::reconnect()
+{
+  return (conn ? conn->close() : seastar::now()).then([this] {
+    if (!mgrmap.get_available()) {
+      logger().warn("No active mgr available yet");
+      return seastar::now();
+    }
+    auto peer = mgrmap.get_active_addrs().legacy_addr();
+    return msgr.connect(peer, CEPH_ENTITY_TYPE_MGR).then(
+      [this](auto xconn) {
+        conn = xconn->release();
+        // ask for the mgrconfigure message
+        auto mgr_open = new MMgrOpen;
```

Comment:
because i need to assign to `MMgrOpen::daemon_name` , but don't really want to down cast it to `MMgrOpen` for this assignment. a better approach is to add a constructor for `MMgrOpen` and only use it for creating `MMgrOpen` instances. as `daemon_name` is not optional. but probably we can do this in a separated PR.